### PR TITLE
Refactor client

### LINKED ISSUE #34: How does the cli know, if the  context is cancelled?

### PR DESCRIPTION
**Pull request by Kyrne**
added tests for the existing rpc calls
fixed some bug in findNode loop
fixed context cancelling in cli loop, closes #34

### CODE REVIEW

File: internal/kademlia/kademliaid/kademliaid.go [hunk 1]
```diff
@@ -11,17 +12,13 @@ import (
 type KademliaID [env.IDLength]byte
 
 // NewKademliaID returns a new KademliaID based on the string input
-func NewKademliaID(data string) KademliaID {
-	// TODO: data is stored as hex at the moment
-	// byte and error
-	decoded, _ := hex.DecodeString(data)
+// TODO: Does not perform length checking yet!
```
Comment (maxluetkemeyer) [inline internal/kademlia/kademliaid/kademliaid.go:15]: If we return an error here, we need to fix a lot of code.

File: internal/network/network.go [hunk 2]
```diff
@@ -3,12 +3,13 @@ package network
...
 	"d7024e_group04/internal/kademlia/contact"
+	"d7024e_group04/internal/kademlia/kademliaid"
 	pb "d7024e_group04/proto"
 )
 
 type ClientRPC interface {
-	SendPing(ctx context.Context, me *contact.Contact, target string) (*contact.Contact, error)
-	SendFindNode(ctx context.Context, me, target *contact.Contact) ([]contact.Contact, error)
+	SendPing(ctx context.Context, me contact.Contact, target string) (*contact.Contact, error)
+	SendFindNode(ctx context.Context, me contact.Contact, candidateAddress string, targetID kademliaid.KademliaID) ([]contact.Contact, error)
```
Comment (maxluetkemeyer) [inline internal/network/network.go:12]: Why did u removed the contact?

File: internal/node/findNode.go [hunk 3]
```diff
@@ -74,6 +74,7 @@ func (n *Node) findNode(rootCtx context.Context, target *contact.Contact) []cont
 				contact.CalcDistance(target.ID)
 
 				// TODO refactor ifs
+				// TODO BUG: FIX DUPLICATE NODES!!!
```
Comment (maxluetkemeyer) [inline internal/node/findNode.go:77]: what does that mean?

File: internal/node/node.go [hunk 4]
```diff
@@ -130,7 +130,7 @@ func (n *Node) GetObject(rootCtx context.Context, hash string) (data string, err
 // TODO: add concurrency
 func (n *Node) pingContacts(ctx context.Context, me *contact.Contact, targetIps []string) (*contact.Contact, error) {
 	for _, targetIp := range targetIps {
-		contact, err := n.Client.SendPing(ctx, me, targetIp+":50051") // TODO fix this port
```
Comment (maxluetkemeyer) [inline internal/node/node.go:133]: can we change targetIps in the func header? The function is called "pingContacts", so i expect to pass multiple contacts and not ip addresses. Also, we then dont need to fix the port, but can easily use the contact address, which includes ip+port.

File: internal/client/client.go [hunk 5]
```diff
@@ -30,33 +29,33 @@ func NewClient(opts ...grpc.DialOption) *Client {
...
+	return nil
 }
 
-// SendPingMessage sends an rpc call to the target contact. If a reply is received the bucket is updated with the target contact.
-func (c *Client) SendPing(ctx context.Context, me *contact.Contact, target string) (*contact.Contact, error) {
-	conn, err := c.NewConnection(target)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create connection to address: %v, err: %v", target, err)
+// SendPingMessage sends a Ping rpc call to the target. Returns the contact of target node.
+func (c *Client) SendPing(ctx context.Context, me contact.Contact, target string) (*contact.Contact, error) {
```
Comment (maxluetkemeyer) [inline internal/client/client.go:46]: What is target? Ip? Ip+Port? KademliaID?

File: internal/client/client.go [hunk 6]
```diff
@@ -67,24 +66,22 @@ func (c *Client) SendPing(ctx context.Context, me *contact.Contact, target strin
 	return &contact, nil
 }
 
-func (c *Client) SendFindNode(ctx context.Context, me, target *contact.Contact) ([]contact.Contact, error) {
-	conn, err := c.NewConnection(target.Address)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create connection to address: %v, err: %v", target.Address, err)
+// SendFindNode sends a FindNode rpc call to the candidate node. Returns the closest nodes to target node from the candidate.
+func (c *Client) SendFindNode(ctx context.Context, me contact.Contact, candidateAddress string, targetID kademliaid.KademliaID) ([]contact.Contact, error) {
```
Comment (maxluetkemeyer) [inline internal/client/client.go:70]: Can't we call it targetContact?

**Comment by hugocornago-ltu** [comment]
We could add the `me` field to the Client structure, as it's being used in every function of Client

**Pull request merged**